### ISSUE TITLE
Game causes OS wide audio stutters mainly during the early parts of a game since the big new game engine update.

### ISSUE THREAD

**Issue by Corosauce** [beyond-all-reason/Beyond-All-Reason#1340]
Since the release of the big game engine change, the game is affecting audio OS wide by causing noticeable stutters, though it appears to go away after a few minutes into a new match.

Observations:
- It only happens when the game is unpaused.
- It did not happen in the older versions of the game engine before this big update that came out a week or 2 ago.
- Doesn't always happen, took a couple games to catch it happening
- Affects youtube audio, vlc audio, foobar2000 audio
- The actual game audio might be fine, not certain

Example of the issue: [](https://www.youtube.com/watch?v=VWsrN0I4us8) - you can hear frequent micro stutters in the audio especially near the end of this video.
- Playing white noise and youtube audio in this clip, you can hear both stuttering

Engine version: 105.1.1-1214-gcf8d087

OS:
Edition	Windows 10 Pro
Version	21H2
Installed on	‎7/‎6/‎2022
OS build	19044.1889
Experience	Windows Feature Experience Pack 120.2212.4180.0

Hardware:
Processor	Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz   4.00 GHz
GPU: AMD 6800 XT
Installed RAM	32.0 GB
Audio device: Trond USB audio adapter, eg: https://www.amazon.com/TROND-Headphone-Microphone-Plug-N-Play-Compatible/dp/B014ANW4VU

**Comment by lhog** [comment]
Always include infolog when you open tickets like that.

**Comment by Corosauce** [comment]
> Always include infolog when you open tickets like that.

My bad, attached the infolog now.

**Comment by Corosauce** [comment]
> Make sure you not running out of RAM and DPC latency is ok

Good call, can confirm DPC latency is spiking like crazy when the game is active, heres what it looks like once a game loads in, left side flat 1000us is before it has loaded in, hearing audio stutters during each red spike:
![dpclat_c2eifDaVgc](https://user-images.githubusercontent.com/1238621/194724252-890ef5ec-5fdb-4803-9bd4-9364d094bf58.png)

Also FYI Memory usage was 16GB used out of 32GB.

**Comment by lhog** [comment]
Can you bisect the issue between the currently running game version (`105.1.1-1212-g94b38a9`) and the version we used before (`105.1.1-966-g597222f`)?
Good strategy is to pick something in the middle of the range, test it to see if stutters are present and then move to the "left" or "right" half of the range depending on the test outcome.

Use https://bucket.p2004a.com/spring-engine-builds/index.html to download engine releases for testing

**Comment by Corosauce** [comment]
> Can you bisect the issue between the currently running game version (`105.1.1-1212-g94b38a9`) and the version we used before (`105.1.1-966-g597222f`)? Good strategy is to pick something in the middle of the range, test it to see if stutters are present and then move to the "left" or "right" half of the range depending on the test outcome.
> 
> Use https://bucket.p2004a.com/spring-engine-builds/index.html to download engine releases for testing

That was a huge help thank you, figured out how to run them with the same params the launcher gives them too.

Decided to just attack the problem as you suggested, same method for minecraft mods to find the problem mod causing unloggable crashes xD, split the list each time you test etc.

Ive found the exact commit that causes the issue, its this, aka build 1000, aka engine_windows64_105.1.1-1000-g128c68a: https://github.com/beyond-all-reason/spring/commit/128c68a78eabd70fef020be778fc679cbbfbd36e

Unless my testing was being intermittent on me, I am 100% sure it is this commit, the previous commit is build 999, and that build worked fine for me, no stutters.

I don't do much c++ anymore, mostly java, but I'm seeing a lot of resolution related changes in there, it must have broken something, maybe something specific to radeon cards.

Bonus discovery, I opened an issue about bizarre multi monitor flickering here: https://github.com/beyond-all-reason/Beyond-All-Reason/issues/1341 . This issue also goes away the same as the audio stuttering, fine on build 999, broken on build 1000.

Just noticed its your commit too @lhog , so I guess I've got the right person on the job xD. Thanks for all the help so far.

**Comment by lhog** [comment]
Hmm... Are you sure about engine_windows64_105.1.1-1000-g128c68a: https://github.com/beyond-all-reason/spring/commit/128c68a78eabd70fef020be778fc679cbbfbd36e ? It looks pretty benign. The commit indeed changed some details of how the window size is fit into display boundaries, but this affecting sound is pretty weird.

A few questions:

- Did your window resolution got changed when switching (back or forth) between 999 and 1000?
- Did you observe any FPS change between the two version?
- I see you play in borderless, yet windowed mode with 2560x1440 resolution, this is also the exact resolution of your first display. Is there a reason you don't go fullscreen?

P.S. Interesting finding on https://github.com/beyond-all-reason/Beyond-All-Reason/issues/1341 Can you attach two infologs:
One named infolog_999.txt, run on 999 engine; another named infolog_1000.txt, run on 1000 engine. In both cases do actions that reproduce glitches in engine 1000 (and do same on 999).

**Comment by Corosauce** [comment]
Finally got some time to get back to this. Some developments, both bad and good.

1. You were right to question if that commit would change anything, once I had finally come back to this to try and do more testing for you the issue started happening in build 999, and build 966 which I had been using reliably for a while before I posted these issues, so my testing methodology was flawed. Sorry If I wasted your time there, frustrating all round. This pretty much invalidates the progress in debugging we've had so far.

2. Your questions about what windowed mode I was using lead me to fiddle around with the options more. So in game you are presented with 3 options, Fullscreen, Borderless, and various resolution options (and then it repeats the options for each monitor). I had not realized the third option were actually fully windowed mode resolutions, not that intuitive from how it presents it ingame, maybe its just me, still silly that I had no tried these options until now. Turns out using fully windowed mode completely removes the audio stuttering and multi monitor glitching out issue in all version of the game.

To reiterate behavior of the resolution options:

Fullscreen: audio stutters, multi monitor glitching
Borderless: audio stutters, multi monitor glitching (behaves 100% exactly like fullscreen)
"2560x1440": perfect experience

For the third option I have to unmaximize and remaximize to get it to be cleanly maximized to the monitor with windows taskbar visible etc, and thats fine. So I'm very happy with this option. For an RTS its fine to not have a fullscreen experience, if this were an FPS things would be different for immersion reasons.

To answer your third question which is still a bit relevant, for lots of FPS games, using borderless is very helpful for rapid alt-tabbing out of the game to another window or monitor without the game taking so long to disengage fullscreen mode, handy for a distracted multi tasking lifestyle, fullscreen is way clunkier in most games and I avoid using it where possible. Though worth noting some games do get better performance in proper fullscreen rather than borderless, so it depends on the game.

Not sure where to go at this point, but I can play happily now just using windowed mode

Given this new info, if you want me to test any versions from here let me know.

We can also just speculate that AMD cards have issues and close this issue as well if thats preferred xD

**Comment by hci2** [comment]
I am experiencing the same issue with the "Borderless Window" setting that is turning my screen shortly black when clicking outside of the current screen where the game runs. It is very useful in a multi-screen setup to use this setting. I am having a Lenovo Legion 5 laptop with NVIDIA GeForce RTX 3070 (GPU) and AMD Ryzen 7 5800H (CPU).

Game: BAR-22966-9651f5e
Engine: 105.1.1-1544-g058c8ea105
Chobby: 2229-3d5e778

Hopefully, this helps you somehow for finding this annoying bug and a solution for it.

**Comment by SethDGamre** [comment]
Calling for a review session on this one because: A. It's old but never got resolved. It's well documented so I don't want to close it for being old.  B. If this was to be fixed, would it be engine side? What's the best way to proceed?

**Comment by SethDGamre** [comment]
I'd like to combine this other issue from this same OP into this one. https://github.com/beyond-all-reason/Beyond-All-Reason/issues/1341 as they share the same game engine version and both glitches are related to screen sizing.

**Comment by WatchTheFort** [comment]
@Corosauce Is this issue still occurring?
